wallet, test: Replace MockableDatabase with in-memory SQLiteDatabase

achow101 opened the pull request:
`MockableDatabase` was introduced for the tests to avoid tying non-database tests to a particular database type. However, since the only database type now is sqlite, and because the mockable behavior is no longer used by the tests, we can replace usage of the `MockabeDatabase` with a SQLite database that lives only in memory.

This is particularly useful for future work that has the wallet make use of SQLite's capabilities more, which are less conducive to having a separate mock database implementation.

achow101 committed:
test: Make duplicating MockableDatabases use cursor and batch

Instead of directly copying the stored records map when duplicating a
MockableDatabase, use a Cursor to read the records, and a Batch to write
them into the new database. This prepares for using SQLite as the
database backend for MockableDatabase.

achow101 committed:
bench, wallet: Make WalletMigration's setup WalletBatch scoped

WalletBatch needs to be in a scope so that it is destroyed before the
database is closed during migration.

PeterWrighten commented:
ACK 78897

rkrux reviewed (commented):
Concept ACK f69852441e56c70edaebeaa7e2db4c6249c7e17c

I have looked in depth only the last commit. Almost all of my suggestions merge towards having child mock class(es) extending the base SQLite specific classes, wherever required. They can be deleted in the future if the need arises without touching the base SQLite classes. Also, I believe they would not make it difficult to achieve the following goal stated in the PR desc.

> This is particularly useful for future work that has the wallet make use of SQLite's capabilities more, which are less conducive to having a separate mock database implementation.

achow101 committed:
wallet: Make Mockable{Database,Batch} subclasses of SQLite classes

The mocking functionality of MockableDatabase, MockableBatch, and
MockableCursor was not really being used. These are changed to be
subclasses of their respective SQLite* classes and will use in-memory
SQLite databases so that the tests are more representative of actual
database behavior.

MockableCursor is removed as there are no overrides needed in
SQLiteCursor for the tests.

achow101 committed:
walletdb: Remove m_mock from SQLiteDatabase

achow101 commented on src/wallet/sqlite.h:
```diff
@@ -95,6 +89,14 @@ class SQLiteBatch : public DatabaseBatch
     bool HasActiveTxn() override { return m_txn; }
+
+    // DO NOT CALL DIRECTLY, use DatabaseBatch::Read, Write, Erase, Exists, and ErasePrefix.
+    // These functions are public for testing only.
```
Made a `MockableSQLiteBatch` and `MockableSQLiteDatabase` which does this.

achow101 commented on src/wallet/sqlite.h:
```diff
@@ -150,12 +152,17 @@ class SQLiteDatabase : public WalletDatabase
+    std::string Format() override
+    {
+        if (m_mock) return "sqlite-mock";
+        return "sqlite";
```
Added `MockableSQLiteDatabase`

achow101 commented on src/wallet/sqlite.h:
```diff
@@ -150,12 +152,17 @@ class SQLiteDatabase : public WalletDatabase
     /** Return paths to all database created files */
     std::vector<fs::path> Files() override
     {
+        if (m_mock) return {};
```
Removed now that there's `MockableSQLiteDatabase`

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3448,7 +3448,7 @@ void CWallet::SetupLegacyScriptPubKeyMan()
     }
 
-    Assert(m_database->Format() == "bdb_ro" || m_database->Format() == "mock");
+    Assert(m_database->Format() == "bdb_ro" || m_database->Format() == "sqlite-mock");
```
I prefer to leave this as-is as it is a more specific check. It checks for the db types that are allowed to be legacy wallets, rather than allowing legacy wallets for all non-sqlite databases.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -345,6 +345,7 @@ bool SQLiteDatabase::Rewrite()
 
 bool SQLiteDatabase::Backup(const std::string& dest) const
 {
+    if (m_mock) return true;
```
Removed now that there's `MockableSQLiteDatabase`

brunoerg reviewed (commented):
Concept ACK

rkrux reviewed (approved):
tACK 1cdb9161774ccf3edcdcb1e482b20be430ed5430

Thanks for incorporating suggestions.

achow101 committed:
wallet, bench: Use TestingSetup in CoinSelection benchmark

sedited reviewed (approved):
ACK b544aabdb373b6e0f50ef7301744e862ada7f525

achow101 commented:
rebased for a silent merge conflict, depends builds were failing.

brunoerg reviewed (approved):
code review ACK 037ea2c714cb1a9a8287a480b87c0ada3aba3ddf

left two comments.

sedited reviewed (approved):
Re-ACK 037ea2c714cb1a9a8287a480b87c0ada3aba3ddf

furszy commented on src/wallet/test/util.cpp:
```diff
@@ -105,7 +105,23 @@ void TestUnloadWallet(std::shared_ptr<CWallet>&& wallet)
+        DataStream key, value;
+        DatabaseCursor::Status status = cursor_orig->Next(key, value);
+        Assert(status != DatabaseCursor::Status::FAIL);
+        if (status != DatabaseCursor::Status::MORE) break;
```
nano nit: should be `status == DatabaseCursor::Status::DONE` (could also do a `while(status != DONE)` to be clearer).

furszy commented on src/wallet/sqlite.cpp:
```diff
@@ -243,13 +247,15 @@ bool SQLiteDatabase::Verify(bilingual_str& error)
-    if (m_mock) {
-        flags |= SQLITE_OPEN_MEMORY; // In memory database for mock db
-    }
+    Open(/*additional_flags*/0);
```
nano nit: a constant "SQLITE_NO_OPEN_FLAG = 0" would be slightly more correct.

furszy reviewed (commented):
Code review ACK 037ea2c714cb1a9a8287a480b87c0ada3aba3ddf
